client/react-dapp/src/App.tsx

```diff
@@ -34,12 +34,26 @@ type Role = {
 };
 
 function App() {
-  const [roles, setRoles] = useState<Role[]>([]);
-  const [did, setDID] = useState<string>("");
+  const userRoles = localStorage.getItem("roles");
+  const did = localStorage.getItem("did");
+  const roles = userRoles ? (JSON.parse(userRoles) as Role[]) : [];
   const [errored, setErrored] = useState<Boolean>(false);
   const [loading, setLoading] = useState<Boolean>(false);
   const [unauthorized, setUnauthorized] = useState<Boolean>(false);
 
+  useEffect(() => {
+    const getRoles = async () => {
+      const res = await axios.get(`${config.backendUrl}/roles`, {
+        withCredentials: true,
+      });
+      if (res.status !== 200) {
+        logout();
+      }
+    };
+
+    getRoles();
+  });
```

Comment:
`useEffect` is executed when the component loads, right? I'm guessing the idea here is that we want to refresh the roles when the page loads. Doesn't this trigger a loop though? If the user isn't logged, then they will `logout()` which will `reload()` the page which will trigger `useEffect` again?

Reply:
yeah. nice catch! you're right, not sure why i didn't encounter the loop behaviour when testing. it's probably not a good idea to reload the page here. perhaps setState might be more appropriate here.

Reply:
Not sure how `setState` is useful in this scenario. Could you make the change or elaborate some more?